[INTERNAL] CustomServerMiddleware: Document karma-ui5 integration caveats (#842)
SAP/ui5-tooling: docs/pages/extensibility/CustomServerMiddleware.md

In this case, a `middlewareUtil` object is provided as a part of the custom middleware's [parameters](#custom-middleware-implementation). Depending on the specification version of the custom middleware, a set of helper functions is available to the implementation. The lowest required specification version for every function is listed in the [MiddlewareUtil API reference](https://sap.github.io/ui5-tooling/v3/api/@ui5_server_middleware_MiddlewareUtil.html).

## Integration with `karma-ui5`

!!! Warning
    **Note:** The Karma project has been [deprecated](https://github.com/karma-runner/karma#karma-is-deprecated-and-is-not-accepting-new-features-or-general-bug-fixes)

[`karma-ui5`](https://github.com/SAP/karma-ui5) is a plugin for the popular [Karma test runner](https://karma-runner.github.io/). Based on your [configuration](https://github.com/SAP/karma-ui5#url), it can fetch resources from a dedicated server or start an internal server using parts of UI5 Tooling.

In case of the latter, any custom middleware configured in the default `ui5.yaml` of the project will be used automatically.

> However, since Karma uses the [`connect`](https://github.com/senchalabs/connect) framework, opposing to the [`express`](https://github.com/expressjs/express) framework used by UI5 Tooling, custom middleware might not always work as expected. `express` is based on `connect` but provides additional functionality. If you plan to use custom middleware in an integrated scenario with karma-ui5, please **restrict the middleware to using `connect` API only**.

Reviewer: This is technically not correct. Connect just provides a middleware framework based on the [native Node.js HTTP request/response interface](https://nodejs.org/api/http.html). Express enhances it with additional functionality, but does not make use of connect. Both just use the native Node.js Server/HTTP functionalities, which is the common ground.

Author: Indeed, the last version that still extended on connect was [Express v3](https://evanhahn.com/understanding-express-3/).

I'll fix that sentence.

Custom middleware defining [Specification Version](../Configuration.md#specification-versions) 2.0 or higher have access to an interface of a [MiddlewareUtil](https://sap.github.io/ui5-tooling/v3/api/@ui5_server_middleware_MiddlewareUtil.html) instance.

In this case, a `middlewareUtil` object is provided as a part of the custom middleware's [parameters](#custom-middleware-implementation). Depending on the specification version of the custom middleware, a set of helper functions is available to the implementation. The lowest required specification version for every function is listed in the [MiddlewareUtil API reference](https://sap.github.io/ui5-tooling/v3/api/@ui5_server_middleware_MiddlewareUtil.html).

## Integration with `karma-ui5`

> !!! Warning
>     **Note:** The Karma project has been [deprecated](https://github.com/karma-runner/karma#karma-is-deprecated-and-is-not-accepting-new-features-or-general-bug-fixes) as of 2023

Reviewer 2: ```suggestion
!!! Warning
    The Karma project has been [deprecated](https://github.com/karma-runner/karma#karma-is-deprecated-and-is-not-accepting-new-features-or-general-bug-fixes) as of 2023
```

Reviewer 2: I'm not terribly familiar with MD, but Warning and Note seem to be distinct entities. Do we still need the **Note** when we have a Warning?

Author: Thanks! I agree, we don't need both here.

[`karma-ui5`](https://github.com/SAP/karma-ui5) is a plugin for the popular [Karma test runner](https://karma-runner.github.io/). Based on your [configuration](https://github.com/SAP/karma-ui5#url), it can fetch resources from a dedicated server or start an internal server using parts of UI5 Tooling.

> In case of the latter, any custom middleware configured in the default `ui5.yaml` of the project will be used automatically.

Reviewer 2: ```suggestion
In the latter case, any custom middleware configured in the default `ui5.yaml` of the project will be used automatically.
```

> However, since Karma uses the [`connect`](https://github.com/senchalabs/connect) framework, as opposed to UI5 Tooling's [`express`](https://github.com/expressjs/express), custom middleware might not always work as expected. Compared to `connect`, the `express` framework provides a greater API to middleware.

Reviewer 2: ```suggestion
However, since Karma uses the [`connect`](https://github.com/senchalabs/connect) framework, as opposed to UI5 Tooling's [`express`](https://github.com/expressjs/express), custom middleware might not always work as expected. Compared to `connect`, the `express` framework provides a more versatile API to middleware.
```

> However, since Karma uses the [`connect`](https://github.com/senchalabs/connect) framework, opposing to [`express`](https://github.com/expressjs/express) used by UI5 Tooling, custom middleware might not always work as expected. Compared to `connect`, the `express` framework provides a greater API to middleware.

Reviewer 2: ```suggestion
However, since Karma uses the [`connect`](https://github.com/senchalabs/connect) framework, as opposed to UI5 Tooling's [`express`](https://github.com/expressjs/express), custom middleware might not always work as expected. Compared to `connect`, the `express` framework provides a greater API to middleware.
```

> However, since Karma uses the [`connect`](https://github.com/senchalabs/connect) framework, opposing to [`express`](https://github.com/expressjs/express) used by UI5 Tooling, custom middleware might not always work as expected. Compared to `connect`, the `express` framework provides a greater API to middleware.

Reviewer 2: "Compared to `connect`, the `express` framework provides a greater API to middleware."

I'm not sure what you mean by "a greater API". Do you mean "more powerful" perhaps?

Author: Technically, express provides more *functions* that can be used. I guess this also makes it more powerful, but I was trying to explain how there are functions in express that do not exist in connect.

Reviewer 2: If express has more functions, I'd expect it to be more powerful. But, just to be on the safe side, we could write 'versatile'.

Author: Alright, thanks 👍

> Therefore, if you plan to use custom middleware in an integrated scenario with `karma-ui5`, you must **restrict the middleware to using [`connect`](https://github.com/senchalabs/connect) API only** to ensure compatibility.

Reviewer 2: ```suggestion
Therefore, if you plan to use custom middleware in an integrated scenario with `karma-ui5`, you must **restrict the middleware to using the [`connect`](https://github.com/senchalabs/connect) API only** to ensure compatibility.
```

> Alternatively, you can start a server with the usual `ui5 serve` command, and [configure the corresponding URL](https://github.com/SAP/karma-ui5#url) for use by `karma-ui5`.

Reviewer 2: ```suggestion
Alternatively, you can start a server with the usual `ui5 serve` command and [configure the corresponding URL](https://github.com/SAP/karma-ui5#url) for use by `karma-ui5`.
```
